### DISCUSSION TITLE
Why Layer's default position is at center of screen when using camera

### DISCUSSION THREAD

**Discussion by johnspeny** [axmolengine/axmol#843] [Q&A]
I am using a yellow LayerColor which is the HUD on which other images are drawn. One of the cameras is supposed to be used for the HUD. The issue is the layer's starting position is at the center of the screen yet it's supposed to be at the bottom left of the screen. The only hack I have tried is using negative values to position it in the bottom left corner. I want it to fill the whole screen but start at the bottom left corner. 

![Camera](https://user-images.githubusercontent.com/30299284/188271881-06f7f5f1-9231-48ed-a155-b0fd9c77dd3a.png)

Here is the code to help you.

Check [Repo](https://github.com/johnspeny/CameraSample)

```cpp

#include "HelloWorldScene.h"

USING_NS_AX;

// on "init" you need to initialize your instance
bool HelloWorld::init()
{
	//////////////////////////////
	// 1. super init first
	if (!Scene::init())
	{
		return false;
	}

	auto visibleSize = Director::getInstance()->getVisibleSize();
	auto origin = Director::getInstance()->getVisibleOrigin();
	auto safeArea = Director::getInstance()->getSafeAreaRect();
	auto safeOrigin = safeArea.origin;

	// default othographic camera for game world
	// user1 camera for HUD
	const auto& winSize = Director::getInstance()->getWinSize();
	_gameCamera = getDefaultCamera();
	_gameCamera->initOrthographic(winSize.width, winSize.height, 1, 1000);
	_gameCamera->setPosition({ 0, 0 });

	// Initialize HUD Camera
	_hudCamera = ax::Camera::createOrthographic(winSize.width, winSize.height, 1, 1000);
	_hudCamera->setDepth(2);
	_hudCamera->setCameraFlag(ax::CameraFlag::USER1);
	const ax::Vec3& eyePosOld = _gameCamera->getPosition3D();
	ax::Vec3 eyePos(eyePosOld);
	_hudCamera->setPosition3D(eyePosOld);
	_hudCamera->lookAt(eyePos);
	_hudCamera->setPosition(0, 0);
	addChild(_hudCamera);

	// Initialize Hud
	_hud = Hud::getInstance();
	_hud->getLayer()->setCameraMask(static_cast<uint16_t>(_hudCamera->getCameraFlag()));
	addChild(_hud->getLayer());

	return true;
}
```

**Comment by DelinWorks** [comment]


**Reply by johnspeny** [reply]
I get wrinkles everytime I face complex issues in code.  I am pretty sure you have 1 or 2 wrinkles already 😆. But thanks for the notice  😄.

**Reply by johnspeny** [reply]
Also check the repo if you need to run it. I have linked it up there

**Reply by DelinWorks** [reply]
> I get wrinkles everytime I face complex issues in code. I am pretty sure you have 1 or 2 wrinkles already 😆. But thanks for the notice 😄.

you tell me 😂

**Reply by DelinWorks** [reply]
> Also check the repo if you need to run it. I have linked it up there

I currently don't have my computer 🙁

**Comment by johnspeny** [comment]
This is the hud.cpp file.

```cpp
#include "Hud.h"

Hud* Hud::getInstance()
{
	static Hud instance;
	return &instance;
}

ax::LayerColor* Hud::getLayer() const
{
	return _layer;
}

Hud::Hud() :
	_layer(ax::LayerColor::create(ax::Color4B::ORANGE)),
	_logo(ax::ui::ImageView::create("CloseNormal.png"))
{
	//_layer->setAnchorPoint({ 0, 0 });
	_layer->addChild(_logo);
}
```

**Reply by DelinWorks** [reply]
A layer color is a 4-vertex (quad) renderable that when created, is set to the screen size and the anchor point of it is 0,0 which is at bottom left of the screen, when you create the camera it's position is 0,0 so you are sitting at the bottom left of the screen which is the reason to why you see a visible half, to fix this you can shift the layer color to half of the negative visible size (as if you were anchoring it to 0.5,0.5 which surprisingly setAnchor doesn't achieve.. likely a bug), and preferably, if you plan on moving the camera, you should attach the layer color to the camera and save the hassle of setting it's position to the camera's, or you can create a component that deals with this.

**Reply by DelinWorks** [reply]
I myself use `clearBackgroundBrush` in the camera class which overrides opengl clear color

**Reply by johnspeny** [reply]
Yah i also tried setting the anchor point to 0.5,0.5 but it had no effect. Then i went on the hack which is shifting the layer using negative values and works. Unfortunately the default position of a box2d body on such a layer even after shifting the layer using the hack seams to be at the center. Weird uhhh

**Reply by johnspeny** [reply]
I have added another layer that acts as the main layer for actual game interactions including physics bodies.

**Reply by johnspeny** [reply]
[https://github.com/johnspeny/CameraSample](https://github.com/johnspeny/CameraSample)

**Comment by umitgoren** [comment]
This happens event without explicitly settings or changing the default camera. Other UI components put into a Panel appear at the middle of  the panel, not bottom-left. So, is this going to be fixed?

**Reply by rh101** [reply]
Have you tried calling `LayerColor::setIgnoreAnchorPointForPosition(false)` in the HUD layer? It should work as you expect it to once that is done, unless it's a completely different issue you're having.

I checked your project, but I'm not sure what it is you're trying to achieve.

Is it possible for you to provide an image of what output you are seeing (if it's different from the screenshot in the initial post), and an image of what output you expect to see?

**Reply by johnspeny** [reply]
I fixed it with a work around. Its already in that repo
```cpp
this->setContentSize(ax::Vec2(ax::Director::getInstance()->getVisibleSize().width, ax::Director::getInstance()->getVisibleSize().height));
this->setIgnoreAnchorPointForPosition(false);
this->setAnchorPoint({ 0.5f, 0.5f });

**Reply by rh101** [reply]
> I fixed it with a work around. Its already in that repo
> 
I understand, which is why I couldn't see the issue when I ran the code in the repo.

**Reply by rh101** [reply]
What's I'm confused about is that you've moved the scene up to that point rather than the HUD layer.  Wasn't the initial issue you were having related to the HUD layer?

**Reply by umitgoren** [reply]
Thanks. I'll try it.

**Reply by rh101** [reply]
```
Hud::Hud() :
	_layer(ax::LayerColor::create(ax::Color4B::ORANGE)),
	_logo(ax::ui::ImageView::create("CloseNormal.png"))
{
	_layer->setContentSize(ax::Vec2(ax::Director::getInstance()->getVisibleSize().width, ax::Director::getInstance()->getVisibleSize().height));
	_layer->setIgnoreAnchorPointForPosition(false);
	_layer->setAnchorPoint(ax::Vec2::ANCHOR_MIDDLE);
	_layer->addChild(_logo);

	_logo->setPosition(ax::Vec2(50, 50));
}
```

Results in this:
![image](https://user-images.githubusercontent.com/8603230/224946023-abbab99b-162b-4eae-9cbb-84b603ed4d44.png)

Just remember to comment out the code in HelloWorld:
```
this->setContentSize(ax::Vec2(ax::Director::getInstance()->getVisibleSize().width, ax::Director::getInstance()->getVisibleSize().height));
//this->setIgnoreAnchorPointForPosition(false);
//this->setAnchorPoint({ 0.5f, 0.5f });
```

**Reply by rh101** [reply]
@johnspeny Did the `setIgnoreAnchorPointForPosition(false);` achieve the desired outcome in the HUD layer?

**Reply by johnspeny** [reply]
@rh101 Will check it when am near a computer. But i like the suggestion of placing it in the layer instead.

**Reply by rh101** [reply]
@johnspeny Just FYI, the code in the linked repo required a few more modifications to get this:
![image](https://user-images.githubusercontent.com/8603230/225018099-04798afb-c93c-4ef1-8769-30af25c67767.png)

Is this the correct output that you are after?

Also, consider making the `GameWorld` and `Hud` classes inherit directly from `ax::Node`, `ax::Layer`, or any other type of node you require, and an instance of those classes can then be directly added to the scene.  You would no longer need to have the `::getLayer()` methods etc., so it simplifies the implementation and usage.

**Reply by johnspeny** [reply]
Yes thats the output needed. Thanks
